RFC: LWC server runtime proposal (#23)
salesforce/lwc-rfcs: text/0112-server-runtime.md

---
title: LWC server runtime
status: DRAFTED
created_at: 2020-01-06
updated_at: 2020-01-06
pr: https://github.com/salesforce/lwc-rfcs/pull/23
---

> # LWC server runtime

Reviewer: @pmdartus "LWC Server Runtime" sounds like an application framework for running LWC, like LWR (and runtime is super overloaded). Can we rename this to "LWC Server Engine" or "LWC DOM Decoupling"

---
title: LWC server runtime
status: DRAFTED
created_at: 2020-01-06
updated_at: 2020-01-06
pr: (leave this empty until the PR is created)
---

# LWC server runtime

## Summary

The LWC engine has been designed from the beginning to run in an environment with access to the DOM APIs. To accommodate server-side rendering (SSR) requirements, we need a way to decouple the engine from the DOM APIs. The existing `@lwc/engine` will be replaced by 3 new packages: 
- `@lwc/engine-core` exposes platform-agnostic APIs and will be used by the different runtime packages to share the common logic.
- `@lwc/engine-dom` exposes LWC APIs available on the browser.

> - `@lwc/engine-server` exposes LWC APIs used for server-side rendering.

Author: Agreed, I wasn't happy with the naming. 

What do you think about `engine-dom-renderer` and `engine-string-renderer`? Even if the package names are more verbose I don't think it's a problem because the only thing developers have to remember if `lwc`.

> - `@lwc/engine-server` exposes LWC APIs used for server-side rendering.

Reviewer 2: I also like the inverted version: `@lwc/client-engine`, `@lwc/server-engine`, `@lwc/engine`

> - `@lwc/engine-server` exposes LWC APIs used for server-side rendering.

Author: I don't think there will be a renaming impact here. The `@lwc/engine` module should never be referenced directly. Userland code should only reference `lwc`, the only place where we might need to change is on the resolver.

> - `@lwc/engine-server` exposes LWC APIs used for server-side rendering.

Reviewer 3: I'm with @ekashida on this, this is NOT about client and server I'm afraid, this is about dom vs dom string. Imagine that in the future we decide to use the dom string output in a client for some reason.

> - `@lwc/engine-server` exposes LWC APIs used for server-side rendering.

Author: > [...] this is about dom vs dom string.

I fully agree with this statement. In this regard, I think it would less confusing if the module that is loaded in a DOM environment contains the term `dom` in its name. I would suggest reaming `@lwc/engine` to `@lwc/engine-dom` in @ekashida proposal.

> - `@lwc/engine-server` exposes LWC APIs used for server-side rendering.

Reviewer 4: > I don't think there will be a renaming impact here. The @lwc/engine module should never be referenced directly.

I was talking about projects that have `@lwc/engine` as a dependency in their `package.json`. I'm assuming that DOM rendering will be the main use case for users so keeping that implicit in the package name seems like it would be less confusing.

## Scope

> LWC SSR support is a broad subject, this proposal only focusses on a subset of it. This proposal covers the following topics:

Reviewer: ```suggestion
LWC SSR support is a broad subject, this proposal only focuses on a subset of it. This proposal covers the following topics:
```

- How to evaluate the existing LWC engine module in a javascript context without DOM access.
- What LWC APIs should be exposed in both environments.

> - How the LWC engine share the same logic when evaluated in both environments.

Reviewer: ```suggestion
- How the LWC engine share the same logic when evaluated in both environments.
```

The following topics are considered out of the scope of this proposal:

> - How the LWC component tree rendered on the server gets serialized and sent to the browser.

Reviewer: By serialized we mean "as HTML", right?

Author: Yes

- How the serialized component tree gets rehydrated when processed by the browser.

## Motivation

The first step in enabling SSR support on LWC is to be able to run the engine in a non-DOM-enabled environment. It is currently impossible to evaluate the engine in such environment due to its hard dependencies on DOM APIs.

While we want to share as much logic as possible between the different environments and make SSR as transparent as possible for component authors, the runtime behavior and APIs exposed by the engine greatly differs between environments. In this regard, it makes sense to break up the current `@lwc/engine` package into multiple packages specifically tailored for each environment.

> Distributing multiple versions of the LWC engine offers the capability to expose different APIs depending on the execution context. On one hand, the current implementation of `createElement` with reaction hooks doesn't make sense in Node.js since there is no DOM tree to attach the created element. In the same way, the `renderToString` method doesn't make sense to ship in a browser.

Reviewer: > with reaction hooks 

that's the key part, right? Since it does make sense to create an element on the server?

Author: The reaction hook is the key part here you are right, reaction hooks is the way we use to get notified when an element is inserted or removed from the DOM. For environments without DOM, hooks are unnecessary.

## Detailed design

The existing `@lwc/engine` will be replaced by 3 new packages:

> - `@lwc/engine-core`: Core logic shared by the different runtime including the rendering engine and the reactivity mechanism. This package should never be consumed directly in an application. This package is agnostic on the underlying rendering medium. The only provide APIs for building custom renderers.

Reviewer: ```suggestion
- `@lwc/engine-core`: Core logic shared by the different runtime including the rendering engine and the reactivity mechanism. This package should never be consumed directly in an application. This package is agnostic on the underlying rendering medium. The only provide APIs for building custom renderers.
```

document.body.appendChild(app);
```

**`server.js`:**

```js
import { createElement, renderToString } from "lwc"; // Aliased to @lwc/engine-server
import App from "c/app";

const app = createElement("c-app", { is: App });
const str = renderToString(app);

console.log(str);
```

### `@lwc/engine-core`

This packages exposes the following platform-agnostic APIs:

- `LightningElement`
- `api`
- `track`
- `readonly`
- `wire`
- `setFeatureFlag`
- `getComponentDef`
- `isComponentConstructor`
- `getComponentConstructor`
- `unwrap`
- `registerTemplate`
- `registerComponent`
- `registerDecorators`
- `sanitizeAttribute`

The DOM APIs used by the rendered engine are injected by the runtime depending on the environment. A list of all the DOM APIs the engine depends upon can be found in the [DOM APIs usage](#dom-apis-usage) section in the Appendix.

### `@lwc/engine-dom`

This package exposes the following APIs:

- `createElement` + reaction hooks
- `buildCustomElementConstructor`
- `isNodeFromTemplate`

This package injects the native DOM APIs into the `@lwc/engine-core` rendering engine.

### `@lwc/engine-server`

This package exposes the following APIs:

> - `createElement(name: string, options: { is: typeof LightningElement }): ServerHTMLElement`: This method creates a new LWC component tree. It follows the same signature as the `createElement` API from `@lwc/engine-dom`. Instead of returning a native `HTMLElement`, this method returns a `ServerHTMLElement` with the public properties, aria reflected properties and HTML global attributed.

Reviewer: ```suggestion
- `createElement(name: string, options: { is: typeof LightningElement }): ServerHTMLElement`: This method creates a new LWC component tree. It follows the same signature as the `createElement` API from `@lwc/engine-dom`. Instead of returning a native `HTMLElement`, this method returns a `ServerHTMLElement` with the public properties, aria reflected properties and HTML global attributes.
```

> - `renderToString(element: ServerHTMLElement): string`: This method synchronously serializes an LWC component tree to HTML. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.

Reviewer 5: How are the components stylesheets passed as part of the string? Are they pre-pended to the string as a <style> tag?

Author: This is one area that will require more investigation. The serialization format is outside the scope of this proposal and will be subject to another RFC: https://github.com/salesforce/lwc-rfcs/blob/ccbd9f38a7ff30d2bd604107e6eb13608366aed2/text/0112-server-runtime.md#scope

Reviewer 5: @seckardt Maybe you want to share your investigation in this area?

Reviewer 6: On our first baby steps towards SSR I didn't reinvent the wheel in that area. I simply reused existing functionality as provided by the `StylesheetFactory`.

> - `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serialize it to string. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.

Reviewer: ```suggestion
- `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serializes it to HTML. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.
```

> - `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serialize it to string. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.

Reviewer 7: How are elements in ShadowDom going to be represented as a string? Is this only compatible with synthetic shadow?

Reviewer 3: Yes, it only works with some sort of synthetic shadow... but that's another RFC... as mentioned in this RFC at the beginning. We have been discussing this with some other folks from google and other companies, we might be able to define this in a way that can be rehydrated on the client side.

> - `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serialize it to string. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.

Reviewer 7: One thing worth exploring: Can we pass `@api` data for the top level component here? If values are known at "tree creation time", we can probably hydrate a good portion of the LWC tree sync as the initial tree is being created

Reviewer 3: this is probably another area to explore in a different RFC since it touches on how much can you render on the server synchronously.

Author: If the data is known before the component is rendered, the current proposal should be sufficient.

```js
import { createElement, renderToString } from "lwc";
import App from "c/app";

const app = createElement("c-app", { is: App });
app.foo = 'Some value';
app.bar = ['Some', 'other', 'value.'];
const str = renderToString(app);

console.log(str);
```

Like a standard `HTMLElement` returned by `createElement` when LWC is running in the browser, the `ServerHTMLElement` return by `created `createElement` on the server would expose:
- the component public properties
- the global HTML properties 
- aria reflection properties

> This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produces a lightweight DOM structure that can be serialized into a string by the `renderToString` method. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.

Reviewer: ```suggestion
This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produces a lightweight DOM structure that can be serialized into a string by the `renderToString` method. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.
```

Reviewer: > Those DOM APIs produce a lightweight DOM structure

Which DOM implementation do you plan to use for that?

Reviewer 3: I think it will be our own bake implementation... We do not plan to allow user-code to talk to the dom, and if they do, we will throw or ignore. The mocks needed here (btw, I will not call them mocks, but implementation) are those APIs needed for the engine to function, so it goes hand to hand.

> This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produces a lightweight DOM structure that can be serialized into a string by the `renderToString` method. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.

Reviewer 7: I'm assuming that "Serialized String" is a string HTML element/tree. Is that correct?

Reviewer 3: Yes, probably we should formalize this as: `A DOMString containing the HTML serialization of the element's descendants. `

> This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produce a lightweight DOM structure. This structure can then be serialized into a string containing the HTML serialization of the element's descendants. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.

Reviewer 5: The POC directly serializes the VDOM, without the need of a DOM structure. Why is a lightweight DOM necessary? Isn't that an extra generation step that would impact the performance?

Author: Yes, this DOM structure will have a performance impact however there needs to be a DOM structure (even really lightweight) to attach the event listeners and dispatch the DOM events.

## Drawbacks

- Requires a complete rewrite of the current `@lwc/engine`

> - Might require substantial changes to the existing tools (jest preset, rollup plugins, ...) to load the right engine depending on the environment.

Reviewer: ```suggestion
- Might require substantial changes to the existing tools (jest preset, rollup plugins, etc.) to load the right engine depending on the environment.
```

## Alternatives

### Using a DOM implementation in JavaScript before evaluating the engine

Prior the evaluation of the LWC engine, we would evaluate a DOM implementation write in JavaScript (like [jsdom](https://github.com/jsdom/jsdom) or [domino](https://github.com/fgnass/domino)). The compelling aspect of this approach is that it requires almost no change to the engine to work.

There are multiple drawbacks with this approach:

- Performance: The engine only relies on a limited set of well-known APIs, leveraging a full DOM implementation for SSR would greatly reduce the SSR performance.

> - Predictability: By attaching the DOM interfaces on the global object, those APIs are not only exposed to the engine but also to the component author. Exposing such APIs to the component author might bring unexpected behavior when component code runs on the server.

Reviewer: I would also add scalability. The mocked APIs would have to be kept in sync with what's actually being used in the engine. Also the engine still wouldn't be reusable in other environments without a corresponding "mock" layer.

Author: I don't think it is the case. There is only a well-defined set of DOM APIs the engine requires to operate. Each runtime (dom and server) is in charge of implementing this contract by passing an implementation of those APIs to the engine.

Reviewer: Maybe it has stabilized by now but when I took that approach I had to add/change multiple mocked APIs during the couple months I touched that code.

Reviewer 3: I don't think that will be that challenging, we will own both packages, and we can keep them in sync.

## Adoption strategy

TBD

> ## How we teach this

Reviewer 8: This would probably be a major version change requiring new documentation.

TBD

## Unresolved questions

**How to load the right version of the engine when resolving the `lwc` identifier?**

For example, when running in Node.js `lwc` might resolve to `@lwc/engine-server` for SSR purposes, but also to `@lwc/engine-dom` when running test via jest and jsdom.

One way to solve this would be to import the platform-specific APIs from the appropriate module: `createElement` from `@lwc/engine-dom`, `createElement` and `renderToString` from `@lwc/engine-server` and use. Platform-agnostic APIs should be imported from `@lwc` (which becomes an alias to `@lwc/engine-core`).

---

## Appendix

### DOM APIs usage

We can break-down the current LWC DOM usages into 3 different categories:

- [DOM constructors used by the engine](#dom-constructors-used-by-the-engine)
- [DOM methods and accessors used by the engine](#dom-methods-and-accessors-used-by-the-engine)

> - [DOM constructors used by component authors](#dom-constructors-used-by-component-authors)

Reviewer 3: component authors might use regular APIs as well, not just constructors

Author: What APIs are you referring to with "regular APIs"? I don't see any other APIs except the one listed in this section that might be involved in SSR.

Reviewer 3: this item is about users code, in their constructor, and in their hook callbacks they can do anything. I believe this bullet point was about that... maybe I'm wrong...what is this item about?

Author: IMO, the only predictable way to make component render on a non-DOM environment is to restrict all the DOM API access when running in such environment. This would force component authors to write more defensive code if they want their components to be eligible for SSR. The same constraints are also enforced by the rest of the popular framework.

For example, if a component needs to attach an event handler on the `window` object in the `connectedCallback` the code needs to check first if the component is running in a browser.

```js
import { LightningElement } from 'lwc';

export default class App extends LightningElement {
  connectedCallback() {
    if (/* check if running in a browser */) {
      window.addEventListener('resize', () => { /* do something */ })
    }
  }
}
```

It's important to note that only a subset of the components written today will be capable to render on the server.

I also commented out the code checking if the component is running in the browser because I don't LWC should provide such mechanism. I would rather delegate this responsibility to the application framework.

Reviewer 3: right, no controversy there... just I don't understand what "DOM constructors used by component authors" means?

Author: Those are DOM interface constructors that developers might need access even in a non-DOM enabled context. So far, I have found 2 constructors that need to be present on the global object: `CustomEvent` and `Event`. 

I am happy to change the title name for something more explicit.

Reviewer 3: Ok, Ok, the important word was "interface", I understand now what you're saying. We are good.
